File: python/93. Restore IP Addresses.py
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        res = []
        self.dfs(s, 0, [], res)
        return res

    def dfs(self, s, index, temp, res):
        if index == 4:
            if not s:
                res.append(".".join(temp))
            return
        for i in range(1, 4):
            if i <= len(s):
                if int(s[:i]) <= 255:
                    self.dfs(s[i:], index + 1, temp + [s[:i]], res)
                if s[0] == "0":
                    break

File: cpp/51-100/93. Restore IP Addresses.cpp
class Solution {
public:
	vector<string> restoreIpAddresses(string s) {
		vector<string> path;
		vector<string> res;
		if(s.size() > 16)
			return res;
		dfs(path, s, res);
		return res;
	}
	void dfs(vector<string> path, string s, vector<string> &res) {
		if (s.empty() && path.size() == 4) {
		    stringstream temp;
			for (int i = 0; i < path.size(); i++) {
				if (i != 0)
					temp << ".";
				temp << path[i];
			}
			res.push_back(temp.str());
		}
		int len = s.size();
		int max_len = min(3, len);
		for (int i = 1; i <= max_len; i++) {
			string str_temp = s.substr(0, i);
			if(str_temp.size()>= 2 && str_temp[0] == '0')
				continue;
			if(stoi(str_temp) > 255)
				continue;
			path.push_back(str_temp);
			dfs(path, s.substr(i), res);
			path.pop_back();
		}
	}
};